packages/compose/src/hooks/use-state-with-history/index.ts

```diff
@@ -0,0 +1,54 @@
+/**
+ * WordPress dependencies
+ */
+import { createUndoManager } from '@wordpress/undo-manager';
+import { useCallback, useRef, useState, useEffect } from '@wordpress/element';
+
+/**
+ * useState with undo/redo history.
+ *
+ * @param initialValue Initial value.
+ * @return Value, setValue, hasUndo, hasRedo, undo, redo.
+ */
+export default function useStateWithHistory< T >( initialValue: T ) {
+	const manager = useRef( createUndoManager() );
```

Comment:
Even though it's an initial value, it's creating a new manager on every render.

Reply:
yeah, I could use an effect but it feels weird, I wish the initial value of useRef was a function.

Reply:
`useMemo`?